examples/computer_vision/deformabledetr_coco_pytorch/data.py

```diff
@@ -0,0 +1,123 @@
+
+from ddetr.util.misc import nested_tensor_from_tensor_list
+from ddetr.datasets.coco import ConvertCocoPolysToMask, make_coco_transforms
+
+
+def unwrap_collate_fn(batch):
+    batch = list(zip(*batch))
+    return batch
+
+
+class GCSBackend:
+    def __init__(self, bucket_name):
+        self._storage_client = storage.Client(project="determined-ai")
+        self._bucket = self._storage_client.bucket(bucket_name)
+
+    def convert_filepath(self, filepath):
+        tokens = filepath.split("/")
+        directory = tokens[-2]
+        filename = tokens[-1]
+        return "{}/{}".format(directory, filename)
+
+    def get(self, filepath):
+        filepath = self.convert_filepath(filepath)
+        blob = self._bucket.blob(filepath)
+        img_str = download_gcs_blob_with_backoff(blob)
+        return img_str
+
+
+class FakeBackend:
+    def __init__(self):
+        self.data = None
+
+    def get(self, filepath):
+        if self.data is None:
+            # Use train_curves.png image as fake data.
+            with open("imgs/train_curves.png", "rb") as f:
+                img_str = f.read()
+            self.data = img_str
+        return self.data
+
+
+class LocalBackend:
+    """
+    This class will load data from harddrive.
+    COCO dataset will be downloaded from source in model_def.py if
+    local backend is specified.
+    """
+
+    def __init__(self, outdir):
+        assert os.path.isdir(outdir)
+        self.outdir = outdir
+
+    def get(self, filepath):
+        # Use train_curves.png image as fake data.
+        with open(os.path.join(self.outdir, filepath), "rb") as f:
+            img_str = f.read()
+        return img_str
+
+
+class CocoDetection(torchvision.datasets.CocoDetection):
+    def __init__(
+        self, backend, root_dir, img_folder, ann_file, transforms, return_masks
+    ):
+        super(CocoDetection, self).__init__(img_folder, ann_file)
+        self.img_folder = img_folder
+        self._transforms = transforms
+        self.prepare = ConvertCocoPolysToMask(return_masks)
+        if backend == "fake":
+            self.backend = FakeBackend()
+        elif backend == "local":
+            self.backend = LocalBackend(img_folder)
+        elif backend == "gcs":
+            self.backend = GCSBackend(root_dir)
+
+    def __getitem__(self, idx):
+        coco = self.coco
+        img_id = self.ids[idx]
+        ann_ids = coco.getAnnIds(imgIds=img_id)
+        target = coco.loadAnns(ann_ids)
+        path = coco.loadImgs(img_id)[0]["file_name"]
+        img_bytes = BytesIO(self.backend.get(os.path.join(self.img_folder, path)))
+
+        img = Image.open(img_bytes).convert("RGB")
+
+        image_id = self.ids[idx]
+        target = {"image_id": image_id, "annotations": target}
+        img, target = self.prepare(img, target)
+        if self._transforms is not None:
+            img, target = self._transforms(img, target)
+        return img, target
+
+
+def build_dataset(image_set, args):
```

Comment:
Is this function intentionally outside of the CocoDetection class?

Reply:
yes, this is to match the organization in the original codebase